### PR TITLE
Validate memory page limits

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: lib/fizzy/parser.cpp [hunk 1]
```diff
@@ -213,6 +213,8 @@ inline parser_result<Memory> parse(const uint8_t* pos, const uint8_t* end)
 {
     Limits limits;
     std::tie(limits, pos) = parse_limits(pos, end);
+    if (limits.min > 65536 || (limits.max.has_value() && *limits.max > 65536))
```
Comment (axic) [inline lib/fizzy/parser.cpp:216]: Should I add a constant into `limits.h`?

File: lib/fizzy/parser.cpp [hunk 2]
```diff
@@ -213,6 +213,8 @@ inline parser_result<Memory> parse(const uint8_t* pos, const uint8_t* end)
...
+        throw validation_error{"maximum memory limit (4GB) exceeded"};
```
Comment (axic) [inline lib/fizzy/parser.cpp:217]: Any message change?

File: lib/fizzy/parser.cpp (hunk 1 above)
Comment (axic) [inline lib/fizzy/parser.cpp:216]: Need to add unit test for this.

Comment (gumb0) [reply lib/fizzy/parser.cpp:216]: There is already `PageSize`

Comment (gumb0) [reply lib/fizzy/parser.cpp:216]: Ah, it's not a page size, then yes, I think it's better to add a new const

Comment (axic) [reply lib/fizzy/parser.cpp:216]: What should it be named?

Comment (gumb0) [reply lib/fizzy/parser.cpp:216]: can be `MemoryPagesValidationLimit`

Comment (axic) [reply lib/fizzy/parser.cpp]: Added.

File: lib/fizzy/parser.cpp (hunk 2 above)
Comment (axic) [reply lib/fizzy/parser.cpp]: Changed it.

File: lib/fizzy/parser.cpp (hunk 1 above)
Comment (axic) [reply lib/fizzy/parser.cpp]: Added unit test.

**Comment by axic** [comment]
Need to squash some commits before merging.

File: lib/fizzy/limits.hpp [hunk 3]
```diff
@@ -6,9 +6,17 @@
 
 namespace fizzy
 {
+// The page size as defined by the WebAssembly 1.0 specification.
 constexpr unsigned PageSize = 65536;
+
+// The maximum memory page limit as defined by the specification. It is only possible to address 4
+// GB (32-bit) of memory.
```
Comment (axic) [inline lib/fizzy/limits.hpp:13]: clang-format splits this up strangely.

**Pull request merged**